### PR TITLE
Samples loading Go built .dll's into C# using P/Invoke

### PR DESCRIPTION
**Pull request by johncburns1**
## Summary

Initial pull request samples loading Go built .dll's into C# using P/Invoke.  The original issue is [here](https://github.com/dotnet/samples/issues/3689).

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/samples?pullRequest=3691) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/samples?pullRequest=3691) before we can accept your contribution.<br/><br/>:x: johncburns1 [sign now](https://cla.dotnetfoundation.org/dotnet/samples?pullRequest=3691)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/samples?pullRequest=3691) it.</sub>

**Comment by BillWagner** [comment]
thanks for submitting this PR @johncburns1 I've looked at the code, and I like what you've built.

Before we merge this, I'm checking with our legal team on the license. The code in this repository is licensed under the MIT license (see https://github.com/dotnet/samples/blob/master/LICENSE-CODE). You've attached the Apache license in this PR. I know those are compatible, but I want to make sure we do things correctly before merging. 

Thanks for your patience.

**Comment by johncburns1** [comment]
If it would be easier, I can just remove the license file headings. I copied from a builder script that I had written in a different repo, but completely changed the entire body of the script. I really just wanted the argument comments on the top of the page and forgot to remove the license info. But I would be happy to just remove it because the code is completely unique to this project.

File: core/interop/godll/cs/README.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+name: "Calling Go functions from C# code"
+urlFragment: "golang-shared-c-dotnet-cs"
+description: "A simple example of using golang's cgo library and .NET's P/Invoke to call Go functions from C# code like a C-style API."
+---
+
+# Calling Go functions from C# code
+
+Go has a library called [cgo](https://golang.org/cmd/cgo/) that enables the creation of Go packages that call C code.  One of cgo's coolest features is that it enables Go packages to be compiled into C shared libraries, exposing Go functions as a C-style APIs.  This sample shows how we can compile a Go project into a Shared C **.dll** and call Go functions from within a .NET Core project using .NET's [P/Invoke](https://docs.microsoft.com/en-us/dotnet/standard/native-interop/pinvoke).
+
+This project consists of 2 primary components:
```
Comment (adegeo) [inline core/interop/godll/cs/README.md:17]:
```suggestion
This project consists of two primary components:
```

File: core/interop/godll/cs/Go/GoBuilder/GoBuilder.csproj [hunk 2]
```diff
@@ -0,0 +1,34 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+    <Configurations>Debug;Release;CodeDoc</Configurations>
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
+    <DefineConstants>DEBUG;TRACE</DefineConstants>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
```
Comment (adegeo) [inline core/interop/godll/cs/Go/GoBuilder/GoBuilder.csproj:10]: Any idea why unsafe is allowed for debug but not for release?

File: core/interop/godll/cs/README.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+1. The .NET Core app that imports and calls functions from the Shared C libraries.  This code is found in `<SolutionDir>/SharedC/Program.cs`.
```
Comment (adegeo) [inline core/interop/godll/cs/README.md:19]: I would remove `<SolutionDir>` and just include the relative path from this readme file. Perhaps you can update the other paths too?

```suggestion
1. The .NET Core app that imports and calls functions from the Shared C libraries.  This code is found in `./SharedC/Program.cs`.
```

Comment (johncburns1) [reply core/interop/godll/cs/README.md:19]: I will fix these.

File: core/interop/godll/cs/Go/GoBuilder/GoBuilder.csproj (hunk 2 above)
Comment (johncburns1) [reply core/interop/godll/cs/Go/GoBuilder/GoBuilder.csproj:10]: I will fix this.

**Comment by adegeo** [comment]
@BillWagner Ping on legal check.

**Comment by BillWagner** [comment]
I got a preliminary response on the legal question and  expect an answer later this week. Thanks for the patience, and for being flexible @johncburns1

**Comment by johncburns1** [comment]
> 
> 
> I got a preliminary response on the legal question and expect an answer later this week. Thanks for the patience, and for being flexible @johncburns1

Do we have an update on the status?

**Comment by BillWagner** [comment]
@johncburns1 

I apologize for the delays. 

We applaud the suggestion here, but we don’t think this is the appropriate place for this sample. This sample seems to have more to do with creating a C export from a Golang project rather than illustrating some .NET feature. DotNet can call C exports through P/Invokes so this specific example doesn’t demonstrate a novel idea. There is also a fair bit of build boiler plate and non-trivial requirements in the example and that isn’t something we feel we should be in charge of maintaining. There is an interesting part about interop with Golang and slices, but that is again an example of how Golang currently implements their interop and not something that Microsoft owns or a pattern we feel we should be endorsing. We don’t feel we should be owning this example which has more to do with exporting and consuming C APIs in Golang rather than anything interesting about a .NET feature set.
 
It is a nice demo and I encourage you to make this available on your personal github or even in a blog post, but not something that should be contained in an official Microsoft owned samples repo.

I'll close this PR now. If you want to revisit it, let us know.

**Pull request closed without merging**